### DISCUSSION TITLE
nebula-graph3.0.0使用if not exists插入边时候出现数据异常

### DISCUSSION THREAD

**Discussion by xiaofaning** [vesoft-inc/nebula#5668] [Q&A]
nebula 插入一条边后，默认match-[e:relation]-(n) 查询会有两个边查询初恋，此时调用insert into edge if not exists 插入相同id指向边的时候更换边的属性字段出现两条不一样的边
![企业微信截图_16915894719240](https://github.com/vesoft-inc/nebula/assets/17213954/f0248cbb-8a37-44ed-9ef6-0d5f6b0038f6)

**Comment by xiaofaning** [comment]
希望各位大佬们能解答一下我这个小菜鸡的问题，万分感谢！！！！！

**Comment by wey-gu** [comment]
What is your query please?

**Reply by xiaofaning** [reply]
After nebula inserts an edge, the default match-[e:relation]-(n) query will have two edge queries. In this case, when you call insert edge if not exists, when the same id is inserted into the edge, two different edges appear in the attribute field of the replacement edge

**Reply by wey-gu** [reply]
`match-[e:relation]-(n)` is not even a valid query

**Reply by xiaofaning** [reply]
match-[e:relation]-(n) return e

**Reply by wey-gu** [reply]
I guess you are querying something like this:

```cypher
match (m)-[e:relation]-(n) return id(m), e, id(n)
```

Now I understand your concern!

**First question, why 2 edges returned.**

This is due to in NebulaGraph edges were duplicated and persisted in two parts of space, one in source, and one in destination.

```
(root@nebula) [basketballplayer]> MATCH (m)-[e:serve]-(n) WHERE src(e) == "player145" RETURN id(m), e, id(n)
+-------------+-----------------------------------------------------------------------+-------------+
| id(m)       | e                                                                     | id(n)       |
+-------------+-----------------------------------------------------------------------+-------------+
| "player145" | [:serve "player145"->"team228" @0 {end_year: 2012, start_year: 2008}] | "team228"   |
| "team228"   | [:serve "player145"->"team228" @0 {end_year: 2012, start_year: 2008}] | "player145" |
...
+-------------+-----------------------------------------------------------------------+-------------+
Got 10 rows (time spent 8.693ms/113.304458ms)
```

Thus, this is recommended(with direction)

```
(root@nebula) [basketballplayer]> MATCH (m)-[e:serve]->(n) WHERE src(e) == "player145" RETURN id(m), e, id(n)
+-------------+-----------------------------------------------------------------------+-----------+
| id(m)       | e                                                                     | id(n)     |
+-------------+-----------------------------------------------------------------------+-----------+
| "player145" | [:serve "player145"->"team228" @0 {end_year: 2012, start_year: 2008}] | "team228" |
...
+-------------+-----------------------------------------------------------------------+-----------+
Got 5 rows (time spent 6.493ms/110.213167ms)

Thu, 10 Aug 2023 12:36:55 CST
```

And further, if you would need to return id of src and dst, this one is even better:
```cypher
MATCH ()-[e:serve]->() WHERE src(e) == "player145" RETURN src(e), properties(e), dst(e)
```

**Second question**, why props are different?

This was due to there is no transaction on this level, thus there is a chance the insert got corruptions, if you are on 3.0.0, you could actually switch on `TOSS` to add transactions on this level.

In your case(which is actually quite rare), you could try redo `INSERT` without the `IF NOT EXISTS` to override the corrupted data :).

Thanks!

**Reply by xiaofaning** [reply]
Thank you very much!
